American Whitewater/AppDelegate.swift

```diff
@@ -0,0 +1,109 @@
+//
+//  AppDelegate.swift
+//  American Whitewater
+//
+//  Updated by David Nelson on 7/4/19.
+//  Copyright © 2019 American Whitewater. All rights reserved.
+//
+
+import UIKit
+import CoreData
+import Firebase
+import OneSignal
+import IQKeyboardManagerSwift
+
+@UIApplicationMain
+class AppDelegate: UIResponder, UIApplicationDelegate {
+
+    var window: UIWindow?
+    
+    // setup CoreData functions and properties
+    lazy var persistentContainer: NSPersistentContainer = {
+        let container = NSPersistentContainer(name: "AmericanWhitewaterV2")
+        container.loadPersistentStores(completionHandler: { (storeDescription, error) in
+            print(storeDescription)
+            if let error = error {
+                fatalError("Unresolved Error: \(error), \(error.localizedDescription)")
+            }
+        })
+        
+        return container
+    }()
+    
+    func saveAWContext() {
+        let context = persistentContainer.viewContext
+        if context.hasChanges {
+            do {
+                try context.save()
+            } catch {
+                let error = error as NSError
+                fatalError("Unresolved error: \(error), \(error.userInfo)")
+            }
+        }
+    }
+
+    func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
+        
+        // setup OneSignal for push notifications
+        // NOTE: only works on actual devices!
+        let onesignalInitSettings = [kOSSettingsKeyAutoPrompt: false]
+
+        // Replace 'YOUR_APP_ID' with your OneSignal App ID.
+        OneSignal.initWithLaunchOptions(launchOptions, appId: "fd159b28-68c6-465e-9301-1bf712f6c435",
+                                        handleNotificationAction: nil, settings: onesignalInitSettings)
+        OneSignal.inFocusDisplayType = OSNotificationDisplayType.notification;
+
+        // Initialize Firebase for crash reporting
+        // and analytics of the app
+        FirebaseApp.configure()
+        
+        // This nice library moves text fields out of the way of the keyboard automagically
+        IQKeyboardManager.shared.enable = true
+        
+        // in case of crash reset the flag saying we are finding reaches
+        DefaultsManager.fetchingreaches = false
+
+        // setting first values for first run
+        if !DefaultsManager.completedFirstRun {
+            DefaultsManager.showRegionFilter = true
```

Comment:
Just wondering, is there a way to have these be the defaults in the DefaultsManager, rather than having to manually set them here?

Reply:
There was an issue where older stored preference data would cause the filtering to crash the app. We've updated the filtering however, I wasn't able to verify if it would crash older apps during the upgrade so I've left this code in there for awhile. We might be able to remove it now? I'm really not sure. It was a hack to get it to stop crashing on upgrades.